### PR TITLE
Add label filtering to CTS

### LINKED ISSUE #7: Implement Selective Testing
`ClusterTestSuite` has defined selectors:
- `matchNames`
- `matchLabels`

Currently, we run all the tests. Implement running only tests defined by selectors.

### PR DESCRIPTION
**Pull request by sjanota**
**Description**

Changes proposed in this pull request:

- `matchLabels` in CTS changed to `matchLabelExpressions` and uses k8s syntax for label filtering
- definition service filters TDs by labels
- definition service ensures that unique TDs are returned 
 - sample added 

**Related issue(s)**

Fixes #7

### CODE REVIEW

File: config/crds/testing_v1alpha1_clustertestsuite.yaml [hunk 1]
```diff
@@ -53,9 +53,10 @@ spec:
               description: Decide which tests to execute. If not provided execute
                 all tests
               properties:
-                matchLabels:
-                  description: Find test definitions by it's labels. TestDefinition
-                    should have AT LEAST one label listed here to be executed.
+                matchLabelExpressions:
+                  description: 'Find test definitions by their labels. TestDefinition
```
Comment (aszecowka) [inline config/crds/testing_v1alpha1_clustertestsuite.yaml:57]: maybe it will be better to point to this documentation: https://kubernetes.io/docs/concepts/overview/working-with-objects/labels ?

File: config/samples/advanced/testsuite-select-by-labels.yaml [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+kind: ClusterTestSuite
+metadata:
+  labels:
+    controller-tools.k8s.io: "1.0"
+  name: testsuite-selected-by-labels
+spec:
+  count: 1
+  selectors:
+    matchLabelExpressions:
+      - "test-duration!=long" # do not execute long running tests
```
Comment (aszecowka) [inline config/samples/advanced/testsuite-select-by-labels.yaml]: Correct me if I am wrong, but in one expression we can provide many conditions that are ANDed:
`environment=production,tier!=frontend`. I assume that if I have many label expressions, they are ORed.

So this example is a little bit confusing because in the first line we ask for executing all tests but ignoring long tests and in the second expression all tests for frontend and backend.
I would rather suggest the following example:
```
- component=frontend,test-duration!=long
- component=backend
```
it means to execute all fast tests for frontend and all tests for the backend.

File: pkg/fetcher/definition.go [hunk 3]
```diff
@@ -22,33 +23,71 @@ type Definition struct {
...
+		return nil, err
 	}
-	// TODO(aszecowka) later so far we return all test definitions for all namespaces (https://github.com/kyma-incubator/octopus/issues/7)
+
+	err = s.findByLabelExpressions(ctx, suite, acc)
+	if err != nil {
+		return nil, err
+	}
+
+	if len(acc) > 0 {
```
Comment (aszecowka) [inline pkg/fetcher/definition.go]: let assume a situation that someone incorrectly defined testsuite with typo in the labels selector. Then, `len(acc)  == 0`, and we fallback to this part of code:
```
	if err := s.reader.List(ctx, &client.ListOptions{Namespace: ""}, &list); err != nil {
		return nil, errors.Wrap(err, "while listing test definitions")
	}
```
which is incorrect IMO.

File: pkg/fetcher/definition.go [hunk 4]
```diff
@@ -22,33 +23,71 @@ type Definition struct {
...
 
+type uniqueTestDefinitions map[types.UID]v1alpha1.TestDefinition
+
 func (s *Definition) FindMatching(suite v1alpha1.ClusterTestSuite) ([]v1alpha1.TestDefinition, error) {
 	ctx := context.TODO()
-	if len(suite.Spec.Selectors.MatchNames) > 0 {
-		return s.findByNames(ctx, suite)
+	acc := make(uniqueTestDefinitions)
+
+	err := s.findByNames(ctx, suite, acc)
```
Comment (aszecowka) [inline pkg/fetcher/definition.go]: I like the previous version (not only because I wrote it ;) ) when findByNames return found testdefinitions. Removing duplicates can happen in this method.

File: pkg/fetcher/definition_test.go [hunk 5]
```diff
@@ -5,6 +5,7 @@ import (
 	"errors"
 	"github.com/kyma-incubator/octopus/pkg/humanerr"
 	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/util/uuid"
```
Comment (aszecowka) [inline pkg/fetcher/definition_test.go]: do we really have to introduce a dependency to a new package? Cannot you mock it, or set it manually:
```UID: "test-a-uid"```?

File: pkg/fetcher/definition_test.go [hunk 6]
```diff
@@ -150,3 +281,20 @@ func (m *mockErrReader) Get(ctx context.Context, key client.ObjectKey, obj runti
 func (m *mockErrReader) List(ctx context.Context, opts *client.ListOptions, list runtime.Object) error {
 	return m.err
 }
+
+type mockListReader struct {
```
Comment (piotrmiskiewicz) [inline pkg/fetcher/definition_test.go:255]: what is the reason to have mockListReader? why do not use directly fakeCli?

File: config/samples/advanced/testsuite-select-by-labels.yaml [hunk 7]
```diff
@@ -0,0 +1,14 @@
...
+apiVersion: testing.kyma-project.io/v1alpha1
+kind: ClusterTestSuite
+metadata:
+  labels:
+    controller-tools.k8s.io: "1.0"
+  name: testsuite-selected-by-labels
+spec:
+  count: 1
+  selectors:
+    matchLabelExpressions:
```
Comment (piotrmiskiewicz) [inline config/samples/advanced/testsuite-select-by-labels.yaml:11]: please, update the documentation about the new field

File: pkg/fetcher/definition_test.go [hunk 8]
```diff
@@ -87,6 +91,133 @@ func TestFindMatching(t *testing.T) {
 
 	})
 
+	t.Run("return tests selected by label expressions", func(t *testing.T) {
```
Comment (aszecowka) [inline pkg/fetcher/definition_test.go:94]: Please add integration test in `testsuite_controller_test.go`

File: pkg/fetcher/definition.go [hunk 9]
```diff
@@ -22,33 +23,71 @@ type Definition struct {
 	reader client.Reader
 }
 
+type uniqueTestDefinitions map[types.UID]v1alpha1.TestDefinition
+
 func (s *Definition) FindMatching(suite v1alpha1.ClusterTestSuite) ([]v1alpha1.TestDefinition, error) {
```
Comment (piotrmiskiewicz) [inline pkg/fetcher/definition.go:26]: you have changed the logic - currently, if no test matches to the `MatchNames` - all tests will be returned. Before your changes - if `MatchNames` is specified and no tests definitions are matching - the method returned empty list.

File: pkg/fetcher/definition_test.go (hunk 6 above)
Comment (aszecowka) [inline pkg/fetcher/definition_test.go:255]: can we get rid of that: https://github.com/kubernetes-sigs/controller-runtime/pull/311 ?

File: pkg/fetcher/definition_test.go [hunk 10]
```diff
@@ -87,6 +91,133 @@ func TestFindMatching(t *testing.T) {
...
+			},
+		})
+		// THEN
+		require.NoError(t, err)
+		assert.Len(t, out, 2)
+		assert.Contains(t, out, *testA)
+		assert.Contains(t, out, *testC)
+	})
+
+	t.Run("return tests returns unique results", func(t *testing.T) {
```
Comment (aszecowka) [inline pkg/fetcher/definition_test.go]: I don't understand this case. We should have for example test definition with 2 labels: 
`tier=backend`, `integration=true` and test suite that have 2 label expressions:
- tier=backend
- integration=true

And test should be listed only once

**Review by aszecowka — CHANGES_REQUESTED** [review]

File: pkg/fetcher/definition.go [hunk 11]
```diff
@@ -22,33 +23,71 @@ type Definition struct {
...
+}
+
+func (s *Definition) findByLabelExpressions(ctx context.Context, suite v1alpha1.ClusterTestSuite, acc uniqueTestDefinitions) error {
+	for _, expr := range suite.Spec.Selectors.MatchLabelExpressions {
+		selector, err := labels.Parse(expr)
+		if err != nil {
+			return errors.Wrapf(err, "while parsing label expression [expression: %s]", expr)
+		}
+		var list v1alpha1.TestDefinitionList
+		if err := s.reader.List(ctx, &client.ListOptions{LabelSelector: selector}, &list); err != nil {
```
Comment (piotrmiskiewicz) [inline pkg/fetcher/definition.go:76]: this line is not tested, I've changed it to: "if err := s.reader.List(ctx, &client.ListOptions{}, &list); err != nil {" and still the test is green. This is important part of the matching logic.
Consider filtering directly by fake client, check the controller runtime version (see https://github.com/kubernetes-sigs/controller-runtime/issues/293)

File: pkg/fetcher/definition_test.go (hunk 8 above)
Comment (piotrmiskiewicz) [inline pkg/fetcher/definition_test.go:94]: Such test is not testing if the matching labels logic works fine, see the comment below line 77. The test is aware of the implementation. I'd like to see a black box test instead of the current one.

File: pkg/fetcher/definition.go [hunk 12]
```diff
@@ -22,33 +23,71 @@ type Definition struct {
...
 	var list v1alpha1.TestDefinitionList
 	if err := s.reader.List(ctx, &client.ListOptions{Namespace: ""}, &list); err != nil {
 		return nil, errors.Wrap(err, "while listing test definitions")
 	}
 	return list.Items, nil
 }
 
-func (s *Definition) findByNames(ctx context.Context, suite v1alpha1.ClusterTestSuite) ([]v1alpha1.TestDefinition, error) {
-	var list []v1alpha1.TestDefinition
+func (s *Definition) findByNames(ctx context.Context, suite v1alpha1.ClusterTestSuite, acc uniqueTestDefinitions) error {
```
Comment (piotrmiskiewicz) [inline pkg/fetcher/definition.go]: why did you move the result to the arg list, for me it is less readable. The name 'findByNames' finds objects and what next? the same comment could be applied to findByLabelExpressions. The code which uses this method is not clear (lines 32 - 37)

File: pkg/fetcher/definition.go [hunk 13]
```diff
@@ -22,33 +23,71 @@ type Definition struct {
...
-		return s.findByNames(ctx, suite)
+	acc := make(uniqueTestDefinitions)
+
+	err := s.findByNames(ctx, suite, acc)
+	if err != nil {
+		return nil, err
 	}
-	// TODO(aszecowka) later so far we return all test definitions for all namespaces (https://github.com/kyma-incubator/octopus/issues/7)
+
+	err = s.findByLabelExpressions(ctx, suite, acc)
```
Comment (piotrmiskiewicz) [inline pkg/fetcher/definition.go]: general question - is it ok to add results of label expressions to the list of tests which matches by name? is it described somewhere in the documentation?

**Review by piotrmiskiewicz — CHANGES_REQUESTED** [review]

File: config/samples/advanced/testsuite-select-by-labels.yaml (hunk 2 above)
Comment (sjanota) [reply config/samples/advanced/testsuite-select-by-labels.yaml]: This example is a lot better, thanks.

File: pkg/fetcher/definition.go (hunk 3 above)
Comment (sjanota) [reply pkg/fetcher/definition.go]: True, haven't thought about it.

File: pkg/fetcher/definition_test.go (hunk 6 above)
Comment (sjanota) [reply pkg/fetcher/definition_test.go:255]: It's available since `controller-runtime` `v0.1.9`, while kubebuilder uses `v0.1.9`.

Comment (sjanota) [reply pkg/fetcher/definition_test.go:255]: That's because fakeCli has a bug and does not filter properly by labels. More info here: kubernetes-sigs/controller-runtime#311

File: pkg/fetcher/definition.go (hunk 13 above)
Comment (aszecowka) [reply pkg/fetcher/definition.go]: I think it is ok, but we have to add that to documentation.

**Comment by aszecowka** [comment]
@sjanota please update Documentation

File: pkg/fetcher/definition.go (hunk 13 above)
Comment (sjanota) [reply pkg/fetcher/definition.go]: It is mentioned in suite docs: 
```
| **spec.selectors** | **NO** | Defines which tests should be executed. You can define tests by specifying their names or labels. Selectors are additive. If not defined, all tests from all Namespaces are executed.
```

File: pkg/fetcher/definition_test.go (hunk 6 above)
Comment (aszecowka) [reply pkg/fetcher/definition_test.go:255]: If we cannot test that with fakeCli we should create "testify" mock

File: pkg/fetcher/definition_test.go (hunk 10 above)
Comment (sjanota) [reply pkg/fetcher/definition_test.go]: The test checks if fetcher returns unique results across all selectors. I'll change it to reflect that more precisely.

File: pkg/fetcher/definition.go (hunk 11 above)
Comment (sjanota) [reply pkg/fetcher/definition.go:76]: It's not available in our controller-runtime version and we cannot use newer as it requires apimachinery 1.13

File: pkg/fetcher/definition_test.go (hunk 6 above)
Comment (sjanota) [reply pkg/fetcher/definition_test.go:255]: @piotrmiskiewicz @aszecowka I chose a different approach. I treat mockListReader as a wrapper around fakeCli and implement the filtering myself. There is more logic in the mock, but then we should be able to switch to fixed fakeCli without changing test scenarios.

File: pkg/fetcher/definition.go (hunk 11 above)
Comment (sjanota) [reply pkg/fetcher/definition.go:76]: I changed the mock to use the LAbelSelector. Now the logic is checked by the test.

File: pkg/fetcher/definition_test.go (hunk 8 above)
Comment (sjanota) [reply pkg/fetcher/definition_test.go:94]: It is changed now.

**Review by piotrmiskiewicz — APPROVED** [review]

File: docs/crd-cluster-test-suite.md [hunk 14]
```diff
@@ -31,7 +31,7 @@ This table lists all the possible parameters of a given resource together with t
 | **metadata.name** |    **YES**   | Specifies the name of the CR. |
 | **spec.selectors** | **NO** | Defines which tests should be executed. You can define tests by specifying their names or labels. Selectors are additive. If not defined, all tests from all Namespaces are executed.
 | **spec.selectors.matchNames** | **NO** | Lists TestDefinitions to execute. For every element on the list, specify **name** and **namespace** that refers to a TestDefinition. |
-| **spec.selectors.matchLabels** | **NO** | Lists labels that match labels of TestDefinitions to execute. A TestDefinition is selected if at least one label matches. This feature is not yet implemented. | 
+| **spec.selectors.matchLabelExpressions** | **NO** | Lists of label expressions that match labels of TestDefinitions to execute. A TestDefinition is selected if at least one label expression matches. For the complete grammar see: https://kubernetes.io/docs/concepts/overview/working-with-objects/labels |
```
Comment (klaudiagrz) [inline docs/crd-cluster-test-suite.md]:
```suggestion
| **spec.selectors.matchLabelExpressions** | **NO** | Lists label expressions that match labels of TestDefinitions to execute. A TestDefinition is selected if at least one label expression matches. See [this](https://kubernetes.io/docs/concepts/overview/working-with-objects/labels) document for more details. | 
```

**Review by klaudiagrz — APPROVED** [review]

**Review by aszecowka — APPROVED** [review]
good job!
